solutions/cf-1528A.mdx

```diff
@@ -0,0 +1,133 @@
+
+<Info title="Why does this work?">
+[Choosing the median minimizes the sum of absolute deviation](https://math.stackexchange.com/questions/85448/why-does-the-median-minimize-ex-c), so we'd want to get as far away from the median as possible.
+</Info>
+
+This observation is all we need. Since setting $u$ to $l_u$ or $r_u$ is always optimal, we'll handle this through DP.
+
+If $\texttt{dp}[i][j]$ stores the maximum sum of beauty values at the subtree rooted at $i$ when $i$'s "beauty type" is $j$. Since vertex $i$ can only take two values, $j$ will be equal to $0$ when $u$'s beauty value is $l_u$ and $1$ when $u$'s beauty value is $r_u$.
+
+Our transitions are:
+
+$\texttt{dp}_{u \space 0} \mathrel{+}= \max(\texttt{dp}_{v \space 0} + |range_{u \space 0} - range_{v \space 0}|, \texttt{dp}_{v \space 1} + |range_{u \space 1} - range_{v \space 0}|)$
+
+$\texttt{dp}_{u \space 1} \mathrel{+}= \max( \texttt{dp}_{v \space 1} + |range_{u \space 1} - range_{v \space 1}|, \texttt{dp}_{v \space 0} + |range_{u \space 0} - range_{v \space 1}|)$
+
+This essentially means:
+
+If $u$ has the beauty $l_u$:
+
+$\texttt{dp}_{u \space 0} \mathrel{+}= \max($beauty if $v$ has a beauty value of $l_v$, beauty if $v$ has a beauty value of $r_v$)
+
+If $u$ has the beauty $r_u$:
+
+$\texttt{dp}_{u \space 1} \mathrel{+}= \max($beauty if $v$ has a beauty value of $l_v$, beauty if $v$ has a beauty value of $r_v$)
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <vector>
+#include <algorithm>
+
+using std::cout;
+using std::endl;
+using std::vector;
+using std::pair;
+
+const int ROOT = 0;
+vector<vector<int>> adj;
+vector<pair<int, int>> range;
+vector<vector<long long>> dp;
+
+// return optimal beauty value for vertex u with subtree v
+long long mx_beauty(int u, int v, bool type) {
+	if (!type) {
+		long long swtch = dp[v][1] + abs(range[u].first - range[v].second);
+		long long sm = dp[v][0] + abs(range[u].first - range[v].first);
+
+		return std::max(swtch, sm);
+	} else {
+		long long swtch = dp[v][1] + abs(range[u].second - range[v].second);
+		long long sm = dp[v][0] + abs(range[u].second - range[v].first);
+
+		return std::max(swtch, sm);
+	}
+}
+void dfs(int u, int par) {
+	for (int v : adj[u]) {
+		if (v == par) {
+			continue;
+		}
+
+		dfs(v, u);
+		dp[u][0] += mx_beauty(u, v, 0);
+		dp[u][1] += mx_beauty(u, v, 1);
+	}
+}
+
+int main() {
+	std::cin.tie(0)->sync_with_stdio(0);
+	
+	int test_num;
+	std::cin >> test_num;
+
+	for (int t = 0; t < test_num; t++) {
+		int n;
+		std::cin >> n;
+
+		range.clear();
+		adj.assign(n, {});
+		dp.assign(n, vector<long long>(2, 0));
+		
+		for (int i = 0; i < n; i++) {
+			int l;
+			int r;
+			std::cin >> l >> r;
+
+			range.emplace_back(std::make_pair(l, r));
```

Comment:
```suggestion
			range.push_back(make_pair(l, r));
```

Reply:
this i agree with

Reply:
not too sure, but if you `emplace_back` doesn't that mean you don't need to `make_pair`? like wouldn't `range.emplace_back(l, r);` just work?

Reply:
oh, i didn't know that, thanks!